Section QA and some reorganisation. (#59)
DanNixon/sine2020_software_guidelines_paper: guidelines_for_collaborative_development_of_sustainable_data_treatment_software.tex

When it is appropriate the idea of the guideline will be abstracted, so that an alternative can be chosen in the future.
Abstracting guidelines in this manner allows them to remain relevant as technology evolves.

\section{Data Formats and Interoperability}
\label{Data Formats and Interoperability}

It is important to use an open and standardised data format, in order to allow users to easily create suitable input files for the software.
A good data format will have open source definitions and documentation, making it easy to understand and to implement file loaders for the software.
It is likely that a library will already exist for loading standardised file formats and should be used whenever possible.

> Any implementation of a standardised data format needs to either follow the standard.

Reviewer: ```suggestion
Any implementation of a standardised data format needs to follow the standard.
```

> The data format should only diverge from the standard if it is absolutely necessary.

Reviewer: ```suggestion
the data format should only diverge from the standard if it is absolutely necessary.
```

Reviewer 2: Why?

Reviewer: I changed the line above to end in and as it flowed better

In the neutron science community NeXus \cite{K_nnecke_2015} has become the default format.
NeXus aims to standardise the way data is represented by different facilities and software by defining a structure on top of the HDF5 file format \cite{HDF5}.
NeXus is a self describing file format where classes can be used to make it easy to understand the data, with no prior knowledge.
Useful metadata (such as instrument configurations and data processing history) is trivial to store in NeXus.

Proprietary file formats should be avoided at all costs.
While there is sometimes an argument for software specific intermediate formats, they should remain intermediate and NeXus should be the preferred interchange format.

Applications may store relevant information (e.g. window layout for a graphical tool) using the facilities provided by the application framework (e.g. \texttt{QSettings} for Qt).

Additional configuration or process setup information should be stored in an easily accessible location.
This data may include version numbers, parameters and filenames.

> This configuration should ideally be in a human readable plain text format, assuming it's data volume does not negate readability or performance.

Reviewer: ```suggestion
The configuration file should be in a human readable plain text format, assuming it's data volume does not negate readability or performance.
```

Popular formats for this purpose include INI, YAML, JSON, TOML and XML, all of which have several library options for common languages and store data as plain text.
Under no circumstances should custom formats be used for configuration files, because there are already a number of well defined data formats for this purpose.

It is vital for the results from scientific software to be reproducible.
However, sometimes there is a need to recreate the results from data treatment, in this case it becomes valuable to be able to replay the series of operations.
Data storage is a concern for most facilities and being able to reproduce data can be beneficial.
An increasingly popular method is to store the unprocessed data and instructions on how to process the data.

> This has the advantage of allowing for multiple sets of instructions with a negligible increase in storage space.

Reviewer: ```suggestion
Therefore, it is possible to record multiple sets of instructions with a negligible increase in storage space.
```

> This means recording a history of each atomic action the software performs along with any relevant numbers/inputs.

Reviewer: swap this with the line above, as it flows better from saying lets record each operation

There are a wide range of cloud Git repository providers, all of which provide similar features.
For a distributed team it is beneficial to use a cloud Git repository (opposed to local hosting on site) to avoid network related bottlenecks.

> \subsection{Build Systems (BS)}

Reviewer: it is used in the text

> \subsection{Build Systems (BS)}

Reviewer 2: Nobody calls a build system BS.
CI and SCM are common abbreviations.
BS is not (at least in this context).

> \subsection{Build Systems (BS)}

Reviewer: Goo point, I just read it as something more explicit

> \subsection{Build Systems (BS)}

Author: 👍

\label{Build systems}

The code base of your software should be managed using an appropriate Build System (BS), which will depend on the choice of programming language(s).
The BS handles the dependency finding and build file generation native to the host platform.
A good choice for C and C++ projects is CMake \cite{CMake}.
This can handle setting up library paths and build tools automatically, allowing developers on any platform to build the software locally.
C and C++ projects can also benefit from use of a dependency manager such as Conan \cite{Conan}.
There are build systems that combine a dependency manager into them.
For Python based projects Python's setuptools is typically sufficient, for specific cases (e.g. complex package dependencies) Conda may be used as an alternative.

\subsection{Quality Assurance (QA)}
\label{Quality assurance}

The quality of the software is one of the most important factors influencing the long-term success of the project.

> Automated testing infrastructure is thus essential in order to provide reliable software product.

Reviewer: ```suggestion
Automated testing infrastructure is essential in order to provide reliable software product.
```

> There is a plethora of testing frameworks available nowadays; almost every modern programming language ships a unit testing framework with its standard libraries, e.g. {\sc CxxTest} for {\sc C++}, {\sc unittest} for {\sc Python}, {\sc JUnit} for {\sc Java}, etc. 

Reviewer: ```suggestion
There is a plethora of testing frameworks available nowadays; almost every modern programming language ships a unit testing framework with its standard libraries (e.g. {\sc CxxTest}, {\sc unittest} and {\sc JUnit} for {\sc C++}, {\sc Python}, and {\sc Java} respectively). 
```

> Moreover, there are third party, more advanced testing frameworks, such as {\sc GoogleTest}, {\sc GoogleMock}, that provide extended techniques for testing, for example, mocking.

Reviewer: ```suggestion
There are more advanced third party testing frameworks (e.g. {\sc GoogleTest}, {\sc GoogleMock}) that provide extended techniques for testing, such as mocking.
```

> In practice, it is almost impossible to achieve 100\% test coverage.

Author: Wasn't aware of that, but if you say it's not widely used in practice, I would abstain from mentioning it.

> In practice, it is almost impossible to achieve 100\% test coverage.

Reviewer 2: It isn't widely used, but I feel that's probably because few people (at least in this community) actually know about it, the benefits and where it should be applied.

However, even in the cases when the tests can not be fully complete, it is a good practice to add a test case each time a bug has been revealed in a certain usage scenario.

> This guarantees that the same misbehaviour will not reoccur in the future.

Reviewer: ```suggestion
This guarantees that the same bug will not reoccur.
```

Apart from unit tests aiming to test the functionality of small units of code in isolation, integration tests are also necessary to ensure that the different parts of the software work together as expected.
Deployment tests on the other hand ensure that the software is properly packaged and can be installed and launched on all the supported platforms.

One peculiarity of experimental data treatment software is that it is particularly hard to test.

> Given the noisiness and stochastic nature of the input data, often empirical methods and heuristics put under the processing logic make it hard to predict the expected output under all possible scenarios.

Reviewer: I am unsure what this is trying to say

Author: I was hoping the next sentence makes it clear, but maybe I am wrong, will think about how to make this clearer.

> For example, an algorithm that is supposed to find a peak may operate as expected if there is a well defined peak in the data, but fail if the statistics for a certain measurement was low.

Reviewer: ```suggestion
For example, an algorithm that is supposed to find a peak may operate as expected if there is a well defined peak in the data, but may fail for low statistics measurements.
```

> Therefore, having unit and integration tests with exactly calculated expected outputs is often not practically feasible.

Reviewer: ```suggestion
Therefore, having unit and integration tests with calculated expected outputs is often not feasible.
```

Reviewer 2: I think the "exactly" is important here, you can test calculated to a certain error.

Reviewer: ok

In such cases reproducibility tests can be used. 

> Those do not, strictly speaking, test the output of the code to be correct, but they ensure that the output remains reproducible over time.

Reviewer: ```suggestion
These tests ensure that the output remains reproducible over time, but do not directly test the output of the code.
```

Reviewer 2: They do test the output of the code, just not it's correctness.

The purpose of these tests is to prevent unexpected changes in the behaviour of the software as a side effect of other development work.
They exploit reference data to compare the output to and the reference itself should be updated only in the case of bug fixes.

Any software that performs computationally heavy operations might need performance tests.
Those are tests that execute a certain aspect of the software over sufficiently large datasets.

> They do not necessarily check the output, but they record the execution time.

Author: I have added a sentence about this.

> Monitoring the evolution of the performance tests timings, one could automatically detect when a newly suggested code change increases significantly the execution time of certain operations.

Reviewer: ```suggestion
Monitoring the evolution of the timings of the performance tests, it is possible to automatically detect when a new code change causes a significant increase in the execution time.
```

Reviewer 2: "performance test results" would do.

When it comes to automated testing, the GUI is arguably the layer that is the hardest to test.

> From the other hand, most often bugs are likely to be found in the graphical interfaces.

Reviewer: ```suggestion
From the other hand, most often bugs are likely to be found in the GUI.
```

However, there are techniques and tools that allow sufficient automation of GUI testing.

> Among those is the separation of the logic into tiers, e.g. Model-View-Presenter (MVP), where the model and the presenter can be covered by unit tests with reciprocal mocking.

Reviewer: ```suggestion
Among those is the separation of the logic into tiers, e.g. Model-View-Presenter (MVP), where the model and the presenter can be covered by unit tests with mocking.
```

> Besides, there are tools available, that allow the test automation of the views themselves.

Reviewer: ```suggestion
There are tools available that allow the automated tests of the views themselves.
```

> While automated testing is indispensable for any software project, the manual (or unscripted) testing should not be undermined.

Reviewer: ```suggestion
While automated testing is indispensable for any software project, the use of manual (or unscripted) testing should not be underestimated. 
```

> Those should be done periodically by the developers or a dedicated QA team, typically before the releases or other milestones.

Reviewer: ```suggestion
These should be done periodically by the developers or a dedicated QA team, typically before the releases or other milestones.
```

The build system handles the dependency finding and build file generation native to the host platform.
A good choice for C and C++ projects is CMake \cite{CMake}.
This can handle setting up library paths and build tools automatically, allowing developers on any platform to build the software locally.
C and C++ projects can also benefit from use of a dependency manager such as Conan \cite{Conan}.
There are build systems that combine a dependency manager into them.
For Python based projects Python's setuptools is typically sufficient, for specific cases (e.g. complex package dependencies) Conda may be used as an alternative.

\subsection{Quality Assurance (QA)}
\label{Quality assurance}

The quality of the software is one of the most important factors influencing the long-term success of the project.
Automated testing infrastructure is essential in order to provide reliable software product.
There is a plethora of testing frameworks available nowadays; almost every modern programming language ships a unit testing framework with its standard libraries (e.g. {\sc unittest} for {\sc Python}, {\sc JUnit} for {\sc Java}). 

> There are more advanced third party testing frameworks, such as {\sc GoogleTest} or {\sc GoogleMock}) that provide extended techniques for testing, for example mocking.

Reviewer 2: I would still mention e.g. pytest here, for the sake of saying you can improve on the languages native tooling.

Reviewer 2: @AnthonyLim23 Could you do this last one before merging?

Reviewer: would adding the sentence:

Some testing frameworks such as  {\sc pytest} can extend/improve native tooling for a framework.

Reviewer 2: Just adding it to the list would be enough.

Author: Just added.

Tools for code reviews are often built into hosted SCM services.
For example, GitHub Pull Requests allow code changes to be viewed and commented on.

\subsection{User and Developer Documentations}
\label{Documentation}

Good documentation will prevent the developers having to spend time explaining how to use and develop the software to users.
The hallmarks of good documentation are that it is easily available, concise and written in a way that is accessible to the target audience.

The simplest and most common form is online documentation, for example Read the Docs \cite{Read_The_Docs} or a wiki page.
The documentation should have a clear structure and be easy to navigate.
An easy way to achieve this is to have a contents page and to make the pages searchable so users can easily find the relevant information, by having tags that summarise the main themes of each page.
The online documentation should be version controlled so it is possible to revert changes and to follow its evolution.

Projects may benefit from having a landing page introducing the project and directing users to commonly used or important resources.

> Examples for this are Hugo \cite{Hugo} and GitHub Pages \cite{GitHub_Pages}.

Reviewer: this sounds like Hugo and GitHub pages are examples of projects with landing pages

Reviewer: ```suggestion
Hugo \cite{Hugo} and GitHub Pages \cite{GitHub_Pages} are examples of resources for creating landing pages.
```

Reviewer 2: tools rather than resources.

Hugo is a powerful and fast static website generator.
It supports themeing and customisation and works with content defined in Markdown.
A free hosting service provided by GitHub Pages, this allows you to commit your website content to a specific branch and have it served by GitHub.
The large benefit here is avoiding the cost and time of maintaining the infrastructure required to host the project's website.

An increasingly popular method of documentation is the video tutorial.
These are extremely useful to users as it allows them to see how the software is used.
For example, video tutorials of intricate GUIs may be beneficial as the user gets to see a complete workflow and there is less chance of steps being undocumented.

> However video tutorials can be costly to create and keep up-to-date with software changes.

Reviewer: ```suggestion
However, video tutorials can be costly to create and keep up-to-date with software changes.
```

It is also valuable to have developer documentation.
This documentation should contain information on the standards used in the code and how to set up a development environment, appropriate tutorials and relevant reference material.

\subsection{Support and Maintenance}
\label{Support and Maintenance}

A key activity for any piece of software is user support.
Unlike documentation this is an interactive link between developers and users.

> As a minimum email support should be provided and users made aware of it.

Reviewer: ```suggestion
As a minimum email support should be provided and advertised to users.
```

The users will be able to download and use the software free of charge, removing a potential barrier in the uptake of the software.
The advantage of an open source license is that it allows anyone to contribute to the code base (but must not bypass any review process) and radically improves the available pool of developers.
Furthermore, in research/academic communities open source software is becoming increasingly compulsory in order to prevent the loss of valuable source code.

\section{Project Management}
\label{Project Management}

Large pieces of software will require some form of management, to ensure that the team are working effectively.
This is often referred to as a governance model.
A governance model will provide a constructive mechanism for deciding the direction of the project and for ensuring it progresses.
This is typically achieved by using project management, which defines a set of guidelines (TODO: cite PFQ https://www.apm.org.uk/book-shop/apm-project-fundamentals-qualification-study-guide/) that will help the project be successful.

> Each project will typically have a sponsor(s) who are ultimately responsible for the project, but will not be involved in its running.

Reviewer: ```suggestion
Each project will typically have a sponsor who are ultimately responsible for the project, but will not be involved in its running.
```

Normally you only have one sponsor

\subsection{Resource Management}
\label{rman}

Resource management can refer to physical equipment, such as computers.
However, in this context it will be used to refer to individuals who are available to work on the project.
The amount of resource available to a project will probably change over time and it is therefore important to plan how best to utilise each resource.
When selecting a resource for a given task it is important to consider their skills/knowledge, the time required to complete the task and the deadline for the task.
If the deadline and time required are similar then it is better to choose a resource who has the relevant skills and experience.
However, the draw back of always choosing the same resource to do similar tasks is that they will become a single point of failure.
If that resource were to leave the team then no-one else would have any knowledge of that part of the project.
Therefore, it is advisable to use tasks as an opportunity to train less experienced/skilled resources when there is sufficient time between the deadline and the expected duration of the task.

\subsection{Stakeholder Engagement}
\label{se}

> Stakeholders is a general term used to describe people who have some interest in your project (e.g. customers, team members and sponsors).

Reviewer: ```suggestion
Stakeholders is a general term used to describe people who have some interest in your project (e.g. customers, team members and sponsor).
```
